SpatialGDK/Source/SpatialGDKEditorToolbar/Private/SpatialGDKEditorToolbar.cpp

```diff
@@ -899,6 +922,163 @@ FString FSpatialGDKEditorToolbarModule::GetOptionalExposedRuntimeIP() const
+
+		return FReply::Unhandled();
+	}
+
+	AddDeploymentTagIfMissing(SpatialConstants::DEV_LOGIN_TAG);
+
+	if (SpatialGDKSettings->IsGenerateSchemaEnabled())
+	{
+		SpatialGDKEditorInstance->GenerateSchema(false);
+	}
+
+	if (SpatialGDKSettings->IsGenerateSnapshotEnabled())
+	{
+		SpatialGDKGenerateSnapshot(GEditor->GetEditorWorldContext().World(), SpatialGDKSettings->GetSpatialOSSnapshotToSave());
+	}
+
+	TSharedRef<FSpatialGDKPackageAssembly> PackageAssembly = SpatialGDKEditorInstance->GetPackageAssemblyRef();
+	PackageAssembly->OnSuccess.BindRaw(this, &FSpatialGDKEditorToolbarModule::OnBuildSuccess);
+	PackageAssembly->BuildAllAndUpload(
+		SpatialGDKSettings->GetAssemblyName(),
+		SpatialGDKSettings->AssemblyBuildConfiguration,
+		TEXT(""),
+		SpatialGDKSettings->bForceAssemblyOverwrite
+	);
+
+	return FReply::Handled();
+}
+
+void FSpatialGDKEditorToolbarModule::OnBuildSuccess()
+{
+	const USpatialGDKEditorSettings* SpatialGDKSettings = GetDefault<USpatialGDKEditorSettings>();
+	if (SpatialGDKSettings->IsSimulatedPlayersEnabled())
+	{
+		IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+		FString BuiltWorkerFolder = GetDefault<USpatialGDKEditorSettings>()->GetBuiltWorkerFolder();
+		FString BuiltSimPlayersName = TEXT("UnrealSimulatedPlayer@Linux.zip");
+		FString BuiltSimPlayerPath = FPaths::Combine(BuiltWorkerFolder, BuiltSimPlayersName);
+
+		if (!PlatformFile.FileExists(*BuiltSimPlayerPath))
+		{
+			FString MissingSimPlayerBuildText = FString::Printf(TEXT("Warning: Detected that %s is missing. To launch a successful SimPlayer deployment ensure that SimPlayers is built and uploaded.\n\nWould you still like to continue with the deployment?"), *BuiltSimPlayersName);
+			EAppReturnType::Type UserAnswer = FMessageDialog::Open(EAppMsgType::YesNo, FText::FromString(MissingSimPlayerBuildText));
+			if (UserAnswer == EAppReturnType::No || UserAnswer == EAppReturnType::Cancel)
+			{
+				OnShowFailedNotification("Failed to launch cloud deployment. SimulatedPlayer archive is missing.");
+				return;
+			}
+		}
+	}
+
+	auto LaunchCloudDeployment = [this]()
+	{
+		SpatialGDKEditorInstance->LaunchCloudDeployment(
+			FSimpleDelegate::CreateLambda([this]()
+			{
+				this->OnShowSuccessNotification("Successfully launched cloud deployment.");
+			}),
+
+			FSimpleDelegate::CreateLambda([this]()
+			{
+				this->OnShowFailedNotification("Failed to launch cloud deployment. See output logs for details.");
+			})
+		);
+	};
+
+	AttemptSpatialAuthResult = Async(EAsyncExecution::Thread, []() { return SpatialCommandUtils::AttemptSpatialAuth(GetDefault<USpatialGDKSettings>()->IsRunningInChina()); },
+		[this, LaunchCloudDeployment]()
+	{
+		if (AttemptSpatialAuthResult.IsReady() && AttemptSpatialAuthResult.Get() == true)
+		{
+			LaunchCloudDeployment();
+		}
+		else
+		{
+			OnShowTaskStartNotification(TEXT("Spatial auth failed attempting to launch cloud deployment."));
```

Comment:
```suggestion
			OnShowTaskStartNotification(TEXT("Failed to authenticate against SpatialOS."));
```
?

Reply:
No, I think we need to say that this happened while launching cloud deployment.

Reply:
`Unable to authenticate. Failed to launch cloud deployment.`
?
Right now I find the sentence confusing as it sounds like spatial auth is launching the cloud deployment

Reply:
Oh I see. Yeah that should work.